internal/mage/engine.go

```diff
@@ -48,3 +52,165 @@ func (t Engine) Lint(ctx context.Context) error {
+
+func (t Engine) Release(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	arches := []string{"amd64", "arm64"}
+	oses := []string{"linux", "darwin"}
+
+	imageRef, err := c.Container().Publish(ctx, engineImageRef, dagger.ContainerPublishOpts{
+		PlatformVariants: engineContainer(c, arches, oses),
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Println("Image published:", imageRef)
+
+	return nil
+}
+
+func localEngine(ctx context.Context, c *dagger.Client) (string, error) {
+	tmpfile, err := os.CreateTemp("", "dagger-engine-export")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile.Name())
+
+	_, err = c.Container().Export(ctx, tmpfile.Name(), dagger.ContainerExportOpts{
+		PlatformVariants: engineContainer(c, []string{runtime.GOARCH}, []string{runtime.GOOS}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	containerName := "test-dagger-engine"
+	volumeName := "test-dagger-engine"
+	imageName := "localhost/test-dagger-engine:latest"
+
+	loadCmd := exec.CommandContext(ctx, "docker", "load", "-i", tmpfile.Name())
+	output, err := loadCmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("docker load failed: %w: %s", err, output)
+	}
+	_, imageID, ok := strings.Cut(string(output), "sha256:")
+	if !ok {
+		return "", fmt.Errorf("unexpected output from docker load: %s", output)
+	}
+	imageID = strings.TrimSpace(imageID)
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"tag",
+		imageID,
+		imageName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker tag: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"rm",
+		"-fv",
+		containerName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker rm: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"run",
+		"-d",
+		"--rm",
+		"-v", volumeName+":/var/lib/buildkit",
+		"--name", containerName,
+		"--privileged",
+		imageName,
+		"--debug",
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker run: %w: %s", err, output)
+	}
+
+	return containerName, nil
+}
+
+func (t Engine) StartLocal(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	containerName, err := localEngine(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("export DAGGER_HOST=docker-container://" + containerName)
+	return nil
+}
+
+func (t Engine) Test(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	containerName, err := localEngine(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	// TODO: run tests inside dagger?
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, "go", "test", "-v", "-count=1", filepath.Join(cwd, "core/integration"))
```

Comment:
Should we continue using `./...`?

With this change we are not longer testing e.g. `./router/merge_test.go`

Reply:
Oh I forgot that there were more tests under router.

Yeah I'll update to do `./...` which will also include running the Go SDK tests against latest engine for now. See my other comment for the longer term idea on how to test each SDK. But there's no harm in `./...` at the moment.

Reply:
> Yeah I'll update to do ./... which will also include running the Go SDK tests against latest engine for now. See my other comment for the longer term idea on how to test each SDK. But there's no harm in ./... at the moment.

Since it's a different go module, `./...` doesn't include that